### PR TITLE
Migrate repoPolicyCheck to OCLIF

### PR DESCRIPTION
**Pull request by RishhiB**
[AB#999](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/999)

### CODE REVIEW

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
+/* eslint-disable unicorn/prefer-module */
+/* eslint-disable import/no-internal-modules */
+/* eslint-disable @typescript-eslint/no-require-imports */
+/* eslint-disable unicorn/import-style */
+/* eslint-disable camelcase */
+/* eslint-disable @typescript-eslint/promise-function-async */
+/* eslint-disable array-callback-return */
+/* eslint-disable @typescript-eslint/strict-boolean-expressions */
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
```
Comment (sonalivdeshpande) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:10]: move this copyright header at the top

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+export class CheckPolicy extends BaseCommand<typeof CheckPolicy.flags> {
+    static description =
+        "Checks that the dependencies between Fluid Framework packages are properly layered.";
+
+    static flags = {
+        resolve: Flags.boolean({
+            description: `Resolve errors if possible`,
+            required: false,
+            char: "h",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:58]:
```suggestion
            char: "f",
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        }),
+        path: Flags.string({
+            description: `Filter file paths by <regex>`,
+            required: false,
+            char: "p",
+        }),
+        stdin: Flags.boolean({
+            description: `Get file from stdin`,
+            required: false,
+            char: "s",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:73]:
```suggestion
```

I'd remove this shortcut since `--stdin` will be a little-used option.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    static flags = {
+        resolve: Flags.boolean({
+            description: `Resolve errors if possible`,
+            required: false,
+            char: "h",
+        }),
+        handler: Flags.string({
+            description: `Filter handler names by <regex>`,
+            required: false,
+            char: "h",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:63]:
```suggestion
            char: "d",
```

Conflicts with `-h` for help.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    });
+};
+
+export class CheckPolicy extends BaseCommand<typeof CheckPolicy.flags> {
+    static description =
+        "Checks that the dependencies between Fluid Framework packages are properly layered.";
+
+    static flags = {
+        resolve: Flags.boolean({
+            description: `Resolve errors if possible`,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:56]:
```suggestion
        fix: Flags.boolean({
            description: `Fix errors if possible`,
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        stdin: Flags.boolean({
+            description: `Get file from stdin`,
+            required: false,
+            char: "s",
+        }),
+        ...BaseCommand.flags,
+    };
+
+    async run() {
+        const { flags } = await this.parse(CheckPolicy);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:79]:
```suggestion
        const flags = this.processedFlags;
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        const exclusions: RegExp[] = require("../../data/exclusions.json").map(
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:80]: Make this path a flag so it can be passed in, then default to `exclusions.json`.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+         * declared file handlers
+         */
+        const handlers: Handler[] = [
+            ...copyrightFileHeaderHandlers,
+            ...npmPackageContentsHandlers,
+            dockerfilePackageHandler,
+            fluidCaseHandler,
+            ...lockfilesHandlers,
+            assertShortCodeHandler,
+        ];
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:100]: You don't need the individual handlers here, so rather than export each of these handlers separately, create and export a new `Handler` array that you can use here. For example:

In the build-tools project:

```ts
export const allHandlers: Handler[] = [
            ...copyrightFileHeaderHandlers,
            ...npmPackageContentsHandlers,
            dockerfilePackageHandler,
            fluidCaseHandler,
            ...lockfilesHandlers,
            assertShortCodeHandler,
        ];
```

Then in build-cli, import `allHandlers.`

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        const handlerActionPerf = new Map<"handle" | "resolve" | "final", Map<string, number>>();
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:101]: Make `"handle" | "resolve" | "final"` a type and use it instead. E.g.

```ts
type policyAction = "handle" | "resolve" | "final";
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        const runPolicyCheck = () => {
+            for (const h of handlers) {
+                const final = h.final;
+                if (final) {
+                    const result = runWithPerf(h.name, "final", () =>
+                        final(pathToGitRoot, flags.resolve),
+                    );
+                    if (result?.error) {
+                        this.exit(1);
+                        this.log(result.error);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:183]: This won't run since exit was called before.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                    }
+
+                    processed++;
+                } else {
+                    this.log(`Excluded: ${line}`);
+                }
+            }
+        };
+
+        const runPolicyCheck = () => {
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:174]: I think this would be better as a pure function that takes a list of handlers as input and throws an exception if a handler fails.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+                    );
+                    if (result?.error) {
+                        this.exit(1);
+                        this.log(result.error);
+                    }
+                }
+            }
+        };
+
+        const logStats = () => {
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:189]: Similar to runPolicyCheck -- this can be a pure function outside of the class that takes handlerActionPerf as input.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+            this.log("Resolving errors if possible.");
+        }
+
+        if (flags.handler !== undefined) {
+            this.log(`Filtering handlers by regex: ${handlerRegex}`);
+        }
+
+        if (flags.path !== undefined) {
+            this.log(`Filtering file paths by regex: ${pathRegex}`);
+        }
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:211]: I think the `run` method will be easier to understand if you move most of the functions aboce this to be defined either on the class or outside the class. Then the run function body is smaller and easier to read and understand end-to-end.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+        if (flags.stdin) {
+            const pipeString = await readStdin();
+
+            if (pipeString) {
+                pipeString.split("\n").map((line: string) => handleLine(line));
+                return;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:218]: Why the return here?

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+            }
+
+            runPolicyCheck();
+            logStats();
+            return;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:223]: Why the return here?

File: build-tools/packages/build-tools/tsconfig.json [hunk 16]
```diff
@@ -15,6 +15,7 @@
         "types": [
             "node"
         ],
+        "esModuleInterop": true,
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/tsconfig.json:18]: Was this needed?

**Review by tylerbutler — CHANGES_REQUESTED** [review]

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+ * Licensed under the MIT License.
+ */
+/* eslint-disable unicorn/prefer-module */
+/* eslint-disable import/no-internal-modules */
+/* eslint-disable @typescript-eslint/no-require-imports */
+/* eslint-disable unicorn/import-style */
+/* eslint-disable camelcase */
+/* eslint-disable @typescript-eslint/promise-function-async */
+/* eslint-disable array-callback-return */
+/* eslint-disable @typescript-eslint/strict-boolean-expressions */
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:12]: Are all of these needed? In general, prefer line-level disables to file-wide.

File: build-tools/packages/build-cli/src/commands/check/policy.ts (hunk 14 above)
Comment (RishhiB) [reply build-tools/packages/build-cli/src/commands/check/policy.ts]: Its not needed, removed

File: build-tools/packages/build-cli/src/commands/check/policy.ts (hunk 15 above)
Comment (RishhiB) [reply build-tools/packages/build-cli/src/commands/check/policy.ts:119]: This return may be needed so the logic without the stdin flag doesnt run after running the logic for the stdin flag

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+    };
+
+    static handlerActionPerf = new Map<policyAction, Map<string, number>>();
+    static processed = 0;
+    static count = 0;
+    static pathToGitRoot = "";
+
+    async run() {
+        const handlerRegex: RegExp =
+            typeof this.processedFlags.handler === "string"
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:77]: This check should always return true because the flag's type is string. Do you mean to check if the flag is undefined?

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+            description: `Filter file paths by <regex>`,
+            required: false,
+            char: "p",
+        }),
+        exclusions: Flags.string({
+            description: `Filter exclusions path`,
+            required: false,
+            char: "e",
+            default: "exclusions.json",
+        }),
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:62]:
```suggestion
        exclusions: Flags.file({
            description: `Path to the exclusions.json file`,
            exists: true,
            required: true,
            char: "e",
            default: "exclusions.json",
        }),
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+                : /.?/;
+        const pathRegex: RegExp =
+            typeof this.processedFlags.path === "string"
+                ? new RegExp(this.processedFlags.path, "i")
+                : /.?/;
+
+        // eslint-disable-next-line @typescript-eslint/no-require-imports, unicorn/prefer-module
+        const exclusions: RegExp[] = require(`../../data/${this.processedFlags.exclusions}`).map(
+            (e: string) => new RegExp(e, "i"),
+        );
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:88]: You need to use the exclusions flag as-is here, not append it to a path. The expectation is that the caller will provide a path to the exclusions file -- as we do when we call policy-check in our scripts -- and that if they do not, we will assume it's in the current working directory, if it's not there, the command will bail early. By changing the flag type to file above, oclif will handle all the checking for us.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/check/policy.ts:88]:
```suggestion
        const exclusions: RegExp[] = require(this.processedFlags.exclusions).map(
            (e: string) => new RegExp(e, "i"),
        );
```
Something like this

File: build-tools/packages/build-cli/src/data/exclusions.json [hunk 21]
```diff
@@ -0,0 +1,7 @@
+[
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/data/exclusions.json:1]: Is this a copy of the existing file, or a move? Either way, I say leave the existing file in place, and then when we update the build scripts in package.json to use the new CLI we can move the file then.

File: build-tools/packages/build-cli/src/commands/check/policy.ts (hunk 18 above)
Comment (RishhiB) [reply build-tools/packages/build-cli/src/commands/check/policy.ts:78]: Yea im using if(typeof this.processedFlags.handler === "string") as a replacement for if(this.processedFlags.handler) since that causes linting errors

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/check/policy.ts:78]:
```suggestion
            this.processedFlags.handler !== undefined
```

Use this instead.

File: build-tools/packages/build-tools/tsconfig.json (hunk 16 above)
Comment (RishhiB) [reply build-tools/packages/build-tools/tsconfig.json]: No it was not needed

File: build-tools/packages/build-tools/src/layerCheck/layerGraph.ts [hunk 22]
```diff
@@ -522,7 +522,7 @@ ${lines.join(newline)}
     }
 
     public static load(root: string, packages: Packages, info?: string): LayerGraph {
-        const layerInfoFile = require(info ?? path.join(__dirname, "..", "..", "data", "layerInfo.json"));
+        const layerInfoFile = require(info ?? path.join(__dirname, "..", "data", "layerInfo.json"));
```
Comment (sonalivdeshpande) [inline build-tools/packages/build-tools/src/layerCheck/layerGraph.ts:525]: why is this change needed?

Comment (RishhiB) [reply build-tools/packages/build-tools/src/layerCheck/layerGraph.ts]: undid this change

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+
+        if (this.processedFlags.path !== undefined) {
+            this.log(`Filtering file paths by regex: ${pathRegex}`);
+        }
+
+        if (this.processedFlags.exclusions === undefined) {
+            this.error("ERROR: No exclusions file provided.");
+        }
+
+        const exclusionsFile = await readJsonAsync(this.processedFlags.exclusions);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:103]: Does readjson return a string? You might need to parse it into an object first.

Comment (RishhiB) [reply build-tools/packages/build-cli/src/commands/check/policy.ts:103]: readJsonAsync seems to return a JSON object, this may throw an error. I think I should wrap it in a try catch.
code for function;
`export async function readJsonAsync(filename: string) {
    const content = await readFileAsync(filename, "utf-8");
    return JSON.parse(content);
}`

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/index.ts [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assertShortCodeHandler, copyrightFileHeaderHandlers, dockerfilePackageHandler, fluidCaseHandler, Handler, lockfilesHandlers, npmPackageContentsHandlers } from "../..";
+
+/**
+ * declared file handlers
+ */
+ export const handlers: Handler[] = [
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/index.ts:11]: nit

```suggestion
 export const policyHandlers: Handler[] = [
```

File: build-tools/packages/build-tools/src/index.ts [hunk 25]
```diff
@@ -33,3 +33,11 @@ export {
 export { LayerGraph } from "./layerCheck/layerGraph";
 export { Timer } from "./common/timer";
 export { execAsync, execWithErrorAsync, readJsonAsync, readFileAsync, writeFileAsync } from "./common/utils";
+export { handler as assertShortCodeHandler } from "./repoPolicyCheck/handlers/assertShortCode";
+export { handlers as copyrightFileHeaderHandlers } from "./repoPolicyCheck/handlers/copyrightFileHeader";
+export { handler as dockerfilePackageHandler } from "./repoPolicyCheck/handlers/dockerfilePackages";
+export { handler as fluidCaseHandler } from "./repoPolicyCheck/handlers/fluidCase";
+export { handlers as lockfilesHandlers } from "./repoPolicyCheck/handlers/lockfiles";
+export { handlers as npmPackageContentsHandlers } from "./repoPolicyCheck/handlers/npmPackages";
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/index.ts:41]: You shouldn't need these exports anymore.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/index.ts [hunk 26]
```diff
@@ -0,0 +1,18 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assertShortCodeHandler, copyrightFileHeaderHandlers, dockerfilePackageHandler, fluidCaseHandler, Handler, lockfilesHandlers, npmPackageContentsHandlers } from "../..";
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/index.ts:6]: import these individually from the source files. e.g.

```ts
import {assertShortCodeHandler} from "./assertShortCode";
```

What you're doing here is creating a new module, and the API for the module is defined by what's exported from this file (index.ts). The individual handlers are "private" to the module, so you import them here and then export them in the shape that you want for your "external/public" API for the module. When you make this change and remove the exports from the root index.ts, you'll effectively "hide" all the individual handlers from the outside world. Even code within the project but outside the module won't be able to access them directly without a lint error. All access will be through the array exported here.

Comment (RishhiB) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/index.ts]: Sorry, I was in the process of doing that

File: build-tools/packages/build-tools/src/index.ts [hunk 27]
```diff
@@ -33,3 +33,5 @@ export {
 export { LayerGraph } from "./layerCheck/layerGraph";
 export { Timer } from "./common/timer";
 export { execAsync, execWithErrorAsync, readJsonAsync, readFileAsync, writeFileAsync } from "./common/utils";
+export { Handler } from "./repoPolicyCheck/common";
+export { policyHandlers } from "./repoPolicyCheck/handlers/index";
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/index.ts:37]:
```suggestion
export { policyHandlers } from "./repoPolicyCheck/handlers";
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 28]
```diff
@@ -0,0 +1,264 @@
...
+            resolve("");
+        }
+    });
+};
+
+type policyAction = "handle" | "resolve" | "final";
+
+export class CheckPolicy extends BaseCommand<typeof CheckPolicy.flags> {
+    static description =
+        "Checks that the dependencies between Fluid Framework packages are following policies.";
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:39]:
```suggestion
        "Checks and applies policies to the files in the repository, such as ensuring a consistent header comment in files, assert tagging, etc.";
```

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+
+        if (stdin.isTTY) {
+            resolve("");
+        }
+    });
+};
+
+type policyAction = "handle" | "resolve" | "final";
+
+export class CheckPolicy extends BaseCommand<typeof CheckPolicy.flags> {
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/check/policy.ts:37]: Can you add some TSDoc comments to the class and functions? Imagine you have to do this again in 6 months -- what do you wish you'd known?

**Review by tylerbutler — APPROVED** [review]
Some nits and stuff, but looks good.

File: build-tools/packages/build-cli/src/commands/check/policy.ts [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+    static processed = 0;
+    static count = 0;
+    static pathToGitRoot = "";
+
+    async run() {
+        const handlerRegex: RegExp =
+            // eslint-disable-next-line no-negated-condition
+            this.processedFlags.handler !== undefined
+                ? new RegExp(this.processedFlags.handler, "i")
+                : /.?/;
```
[the remark this answers is not in the document]
Comment (RishhiB) [reply build-tools/packages/build-cli/src/commands/check/policy.ts:80]: Yea I was thinking the same thing :)

**Comment by RishhiB** [comment]
Thanks for all the help @tylerbutler and @sonalivdeshpande !!
